### PR TITLE
Add `SaveMixin`

### LINKED ISSUE #1002: Add `SaveMixin`
### 🚀 Feature Request

We want to add possibility to save pipelines/models/transforms to disk.

### Proposal

1. Add class into `etna/core/saving`:
```python
class AbstractSaveable(ABC):

	@abstractmethod
     	def save(path: pathlib.Path):
             pass			
	
	@abstractmethod
	@classmethod
	def load(source: pathlib.Path) -> "AbstractSaveable":
             pass	

class SaveMixin(AbstractSaveable):

	def save(path: pathlib.Path):
		with Zipfile(path, "w") as zip_file:
                         # for getting version use: https://stackoverflow.com/a/32965521
                         metadata = {"etna_version": ..., "class": ...}
			with zip_file.open("metadata.json", "w") as output_file:
				json.dumps(output_file, zip_file)
		
			with zip_file.open("object.pkl", "wb") as output_file:
				pickle.dumps(output_file, self)				
	
	@classmethod
	def load(source: pathlib.Path) -> "SaveMixin":
		with Zipfile(source, "r") as zip_file:
			with zip_file.open("metadata.json", "r") as input_file:
				metadata = json.loads(output_file, zip_file)
                         cur_etna_version = ...
                         saved_etna_version = ...
                         # if major version is different then give a warning
			with zip_file.open("obj.pkl", "rb") as input_file:
				return pickle.loads(input_file.read())
```
2. Inherit `Transform` from `SaveMixin`.
3. Inherit `AbstractModel` from `SaveMixin`.

### Test cases

1. Test mixin.
  * Check warning on different version.
3. Test every transform on correct saving/loading.
4. Test every model on correct saving/loading. Models for which it is ok to fail for now:
* catboost-based
* `ProphetModel`
* statsmodels
* old nns
* new nns

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by Mr-Geekman**
## Before submitting (must do checklist)

- [x] Did you read the [contribution guide](https://github.com/tinkoff-ai/etna/blob/master/CONTRIBUTING.md)?
- [x] Did you update the docs? We use Numpy format for all the methods and classes.
- [x] Did you write any new necessary tests?
- [x] Did you update the [CHANGELOG](https://github.com/tinkoff-ai/etna/blob/master/CHANGELOG.md)?

## Proposed Changes
Look at #1002.

## Closing issues
Closes #1002.

### CODE REVIEW

File: etna/transforms/decomposition/stl.py [hunk 1]
```diff
@@ -194,7 +194,9 @@ def __init__(
             2. "holt": ``ETSModel(data, trend='add')``
 
             Custom model should be a subclass of :py:class:`statsmodels.tsa.base.tsa_model.TimeSeriesModel`
-            and have method ``get_prediction`` (not just ``predict``)
+            and have method ``get_prediction`` (not just ``predict``).
+
+            During ``save`` pickle is used and model should support saving and loading with pickle.
```
Comment (martins0n) [inline etna/transforms/decomposition/stl.py:199]: Why do we need to specify this?

File: tests/test_core/test_mixins.py [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+
+        with ZipFile(path, "r") as zip_file:
+            files = zip_file.namelist()
+            assert sorted(files) == ["metadata.json", "object.pkl"]
+
+            with zip_file.open("metadata.json", "r") as input_file:
+                metadata_bytes = input_file.read()
+            metadata_str = metadata_bytes.decode("utf-8")
+            metadata = json.loads(metadata_str)
+            assert sorted(metadata.keys()) == ["class", "etna_version"]
```
Comment (martins0n) [inline tests/test_core/test_mixins.py:41]: Add class name check?

File: tests/test_models/nn/test_tft.py [hunk 3]
```diff
@@ -178,3 +179,12 @@ def test_prediction_interval_run_infuture_warning_loss(example_tsds):
         segment_slice = forecast[:, segment, :][segment]
         assert {"target"}.issubset(segment_slice.columns)
         assert {"target_0.02", "target_0.98"}.isdisjoint(segment_slice.columns)
+
+
+@pytest.mark.xfail(reason="Should be fixed in inference-v2.0")
```
Comment (martins0n) [inline tests/test_models/nn/test_tft.py:184]: strict flag

File: tests/test_models/nn/test_mlp.py [hunk 4]
```diff
@@ -101,3 +102,20 @@ def test_mlp_layers():
         nn.Linear(in_features=3, out_features=10), nn.ReLU(), nn.Linear(in_features=10, out_features=1)
     )
     assert repr(model_) == repr(model.mlp)
+
+
+@pytest.mark.xfail(reason="Working, but should be fixed in inference-v2.0")
```
Comment (martins0n) [inline tests/test_models/nn/test_mlp.py:107]: :thonking:

**Review by martins0n — CHANGES_REQUESTED** [review]

File: etna/transforms/decomposition/stl.py (hunk 1 above)
Comment (Mr-Geekman) [reply etna/transforms/decomposition/stl.py:199]: I think there is a possibility that some users might give some model that doesn't support pickling.

File: etna/core/mixins.py [hunk 5]
```diff
@@ -89,3 +96,83 @@ class StringEnumWithRepr(str, Enum):
...
+        ----------
+        path:
+            Path to save object to.
+        """
+        with ZipFile(path, "w") as zip_file:
+            metadata = {
+                "etna_version": get_etna_version(),
+                "class": inspect.getmodule(self).__name__,  # type: ignore
+            }
+            metadata_str = json.dumps(metadata)
```
[the remark this answers is not in the document]
Comment (Mr-Geekman) [reply etna/core/mixins.py]: Fixed.

File: tests/test_core/test_mixins.py (hunk 2 above)
Comment (Mr-Geekman) [reply tests/test_core/test_mixins.py:41]: Fixed.

File: tests/test_models/nn/test_tft.py (hunk 3 above)
Comment (Mr-Geekman) [reply tests/test_models/nn/test_tft.py]: Fixed.

**Review by martins0n — APPROVED** [review]
👍

**Pull request merged**